rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,418 @@
+challenge containing a nonce. To avoid completely implementing our own
+authentication ceremony, clients can use `go-jose` to marshal and sign a JWT
+which can then be verified easily on the server.
+
+TODO: This needs significant further elaboration and feedback.
+
+#### Client-Side Changes in `tbot`
+
+Bots should be informed of their number of remaining rejoins. We can give bots
+permission to view their own join token, or include the number of remaining
+rejoins as an informational field in the bot's current user certificate. We
+should then expose this as a Prometheus metric to allow for alerting if a bot
+
+We should also investigate hardware key storage backends using HSMs / PKCS#11.
+We have some prior art here in Teleport's [HSM
+support](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/).
+
+#### Non-Terraform UX
+
+TODO
+
+#### Remaining Downsides
+
+- Repairing a bot that has exhausted all of its rejoins is still a semi-manual
+  process. It is significantly easier, and does not necessarily require any
+  changes on the impacted bot node itself, but is still annoying. Users can opt
+  out of this by setting `.spec.rejoining.unlimited=true`, but this has obvious
+  security implications.
+
+- Effort required to configure IaC / Terraform is still fairly high, even if
+  reduced.
+
+### Other Supporting Improvements
+
+Alongside `challenge` joining, we have several UX proposals to further improve
+the usability of non-delegated joining.
+
+#### Longer-Lived Bots
+
+The renewable identity's 24 hour maximum TTL is too restrictive and should be
+lengthened. We propose raising this limit to 7 days, but keeping the default
+(1hr) the same.
+
+#### Bot Instance Locking
+
+When we introduced bot instances in
+[RFD0162](0162-machine-id-token-join-method-bot-instance.md), it allowed many
+bot instances to join under a single bot (and Teleport user). Generation counter
+checks were moved out of a bot's user and into bot instances, however when a
+generation counter mismatch occurs, the resulting lock is still filed against
+the user as a whole. This means a generation counter lockout in one instance can
+easily impact all other instances under the same bot.
+
+We should make this more granular and make locks that can target a bot instance
+UUID, and as this RFD introduces a method for bots to rejoin as a new instance,
+a lock target for specific join tokens. We may need to introduce join token
+tracking, e.g. by introducing a join token certificate field.
+
+#### Token CLI UX Improvements
+
+Instead of separating join method and token name, we propose combining the two
+into a single CLI flag and referring to tokens as a (method, value) tuple. For
+example:
+
+```
+$ tctl bots add example --roles=access
+The bot token: challenge:04f0ceff1bd0589ba45c1832dfc8feaf
+This token will expire in 59 minutes.
+
+[...snip...]
+
+$ tbot start --token=challenge:04f0ceff1bd0589ba45c1832dfc8feaf ...
+```
+
+## Alternatives and Future Extensions
```

Comment:
One thing that I think would be useful to consider within the RFD is whether this process should be tied to a join token or directly to a Bot Instance. I think join token is likely the correct choice but I think worth at least considering.

Reply:
I added a section under alternatives to discuss this. I think the most straightforward argument in favor of keeping state out of bot instances is that it'd make it harder to extend this join method to regular nodes if we ever wanted to do that.